### ISSUE TITLE
Spacing tab displays different result compared to exported font

### ISSUE THREAD

**Issue by ciqven** [johanmattssonm/birdfont#183]
Hello Johan,

Since the last issue of yesterday I tried to figure out what was going wrong, and I have found some things.

For background, I am currently trying to use a mix of ligatures and contextual substitutions to create initial and combining forms of a glyph. Each of these forms can then have combining diacritics. However, it seems the spacing tab inside Birdfont shows the same sequence of characters different from other applications rendering the exported font.

Here is an example of the sequence `U+101220 U+10122A U+101220` in Birdfont:
![font_birdfont](https://github.com/user-attachments/assets/8e4484fa-558e-4191-b528-7e69b7f0b447)

This is what I want, so I consider this correct. However, here is the same character sequence in other programs using the exported font:
GIMP:
![font_gimp](https://github.com/user-attachments/assets/735b3c77-d048-442b-9ef9-0ac276ef9960)

Firefox:
![font_firefox](https://github.com/user-attachments/assets/3f9d3497-4ced-4a32-b0d8-448f9c3ab7e3)

Clearly something is going wrong here, the last character is not getting transformed the same way. I have attached the save file and output (a slimmed down version, no 3K glyphs in this one).

[font.zip](https://github.com/user-attachments/files/17480072/font.zip)

On a slightly related note, could you tell me the order that ligatures and contextual substitutions resolve in? I am currently assuming contextuals get resolved first, but maybe I am mistaken...

**Comment by johanmattssonm** [comment]
I am pretty sure regular ligatures come first when the font is rendered.

Are you using ligatures as input to other ligatures? I am not sure if
that is supported in the font specifications.

I did look at some random examples in your font (see attached html
file). Things seem to work in these examples, I am not sure about the
sequence  U+101220 U+10122A U+101220 and it is pretty hard to figure
out what's happening. What do you want to do for the input U+101220
U+10122A U+101220?

**Comment by johanmattssonm** [comment]
Many thanks for the report. Your hint about the order of substitution could be very important. I am not sure if I found a solution or if I just made things worse but I have made a new test font with some updates to ligatures. See the attached file. It is way to early for a release but I would very much appreciate of you could take a look at the generated OTF file and see if things match better in birdfont and other programs. From my end, it seems there’s been some improvement.

[2024-10-23-Tyeirem-gsub-update.zip](https://github.com/user-attachments/files/17483085/2024-10-23-Tyeirem-gsub-update.zip)

**Comment by johanmattssonm** [comment]
Order of substitution in this case refers to regular ligatures being processed before contextual ligatures.

**Comment by johanmattssonm** [comment]
I can export other versions for you if you have a more complete set of ligature substitutions to test.

**Comment by ciqven** [comment]
Thank you very much for the solution, the font you provided seems to do exactly what I want within the scope of the save/font provided! 

I have attached two more `.birdfont` save files. The first one is a small one, which has the full set of substitutions but limited to 2 vowels and 3 consonants. 

[font_saves.zip](https://github.com/user-attachments/files/17487836/font_saves.zip)

The second file has all the combinations I eventually want, and here you can see another problem: this file is a whopping 35 MB, of which the majority is >30K `<contextual>` lines. This takes about 20 minutes to export, but does complete in the beta version you've sent me.

**Comment by johanmattssonm** [comment]
Thank you. I will try to export it and send you a beta version. I am
not sure what the maximum number of ligatures that can be added to one
font is but I cross my fingers and hope that it is more than 30K.

**Comment by ciqven** [comment]
Thank you for sending me the new beta. I think I have it working now!

In your earlier comment you replied to me that the order was `ligature -> contextual ligature`. In some of my generation code I had been assuming the opposite, which broke lookup in various cases. After quickly reversing this logic it seems to work better now with the limited save file. I am going to do a bit of a refactor on my end and try exporting the full font afterwards, I'll keep you updated.

**Comment by johanmattssonm** [comment]
Brilliant, thanks I will keep things in beta state until you have got
things working.

**Comment by ciqven** [comment]
Hello Johan,

Everything seems to work except for one last problem. I am not sure how to exactly implement this the right way using ligatures and contextuals, so let me attempt illustrate it.

![baseglyphs](https://github.com/user-attachments/assets/30f0bfe9-f53f-450c-ab6d-d86a2a8422fe)
In the font, I have these two glyphs. The first (`U+101220`) is the vowel `i`, the second (`U+10122C`) is a diacritic that changes `i` to `ih`.

![combining](https://github.com/user-attachments/assets/d4339922-7e45-4b02-ab86-ba6c1e29c320)
The vowel `i` has an initial form and a form for when it occurs isolated (not behind any consonants) in the middle/end of a word. I use a contextual substitution here to change it to the `l.i.comb` glyph when something part of a word is in front of it.

![morph](https://github.com/user-attachments/assets/78950978-d994-4ed1-b463-1cff2b7c10a3)
When followed by the morphing diacritic, the vowel `i` turns to `ih` - represented by the `l.ih` glyph.

The above all works great so far, but here is my problem:
![total](https://github.com/user-attachments/assets/f1007bbb-ef19-434d-92b2-f3c8c280785a)
When the morphing diacritic follows a middle/end vowel `i`, it should turn into a morphed middle/end vowel - **not** an initial one. 

The above image has a correct (first, green tint) and incorrect (second, red tint) example. Birdfont's Spacing tab displays the correct example when I enter `U+101220 U+10122C U+101220 U+10122C` (`i morph i morph`). Everything else I tested with the exported font displays the incorrect example.

### Current order
I am currently first applying diacritics to the vowel using ligatures, for example the morphing diacritic on the `i` vowel with the line
```<ligature sequence="􁈠 􁈬" replacement="l.ih" priority="1"/>```

Then after that, I have contextual substitutions that turn the default initial vowels and the created ligatures into middle/end variants:
```<contextual ligature="l.i.comb" backtrack="..." input="􁈠" lookahead="" />
<contextual ligature="l.ih.comb" backtrack="..." input="l.ih" lookahead="" />
```
_(I put `...` as the backtrack value in this example, in reality there is a line like this for every possible codepoint/glyph that is "part of a word")_


So Birdfont itself does actually show my desired outcome in the Spacing tab:
![2024-10-23-153748_241x231_scrot](https://github.com/user-attachments/assets/b147b263-3298-4b80-9352-d8ae3e5d0727)

Is there a way to get this to work in the export as well?

**Comment by ciqven** [comment]
I almost forgot, attached is a demo save and export to test with
[demo.zip](https://github.com/user-attachments/files/17492644/demo.zip)

**Comment by johanmattssonm** [comment]
Thanks, I took a quick look at it now but I need to dig in to this later
tonight.

**Comment by johanmattssonm** [comment]
Hi. I spent some time on this and I believe that this was probably
working due to a bug in the Birdfont kerning tab. A ligature can not
substitute another ligature but a ligature can be added as a part of
the text coming before and after the text that you want to replace in
contextual substitution. I have made some modifications and some
tests. Please let me know if I got anything wrong. I will send you a
new beta.

Thank you for the help when it comes to figuring all this out.

**Comment by ciqven** [comment]
Thanks for the new beta, and for your work in debugging all of this with me. I will give the new beta a try and adapt my ligature generation order when I am back from work tonight.

**Comment by ciqven** [comment]
Hey Johan,

Thanks for the last reply, I managed to get this working on a small scale by using contextual substitutions to add diacritics to those initial vowel forms as well. This seems to work perfectly if I leave both `backtrack` and `lookahead` empty, and fixes the priority issue I had with resolving.

I have been testing the new version, and found a couple more points I am unsure of. The main one is that around roughly 1040 contextual substitutions, ligature lookup completely breaks. I have attached a zip with two saves and their output. Aside from the exported font names, the difference is only one `<contextual ... />`  at the end of the file.

[1linediff.zip](https://github.com/user-attachments/files/17510351/1linediff.zip)

When I load these fonts into Firefox or Gimp and test the string `􁈠􁈪` (`U+101220 U+10122A`), the following happens:
![difference](https://github.com/user-attachments/assets/22cf4a48-fde6-4d8c-8178-727999d975ae)

This change between working and not working still happens when I remove/add another `<contextual />` line in the file, so I seem to be hitting some kind of limit here.

Another issue I found is that in the latest beta you sent me, the string `􁈠􁈪􁈠􁈬` (`U+101220 U+10122A U+101220 U+10122C`) looks different again. 
![ffox_birdfont_diff](https://github.com/user-attachments/assets/592b23b5-01c3-4b4a-b95d-0a494fe51be9)

The Firefox version is what I want, and what I assume is correct. This can be tested with the "Working" version in the ZIP archive I attached to this post.

**Comment by johanmattssonm** [comment]
Thank you, I will look into this over the weekend.

**Comment by johanmattssonm** [comment]
Thanks.

The kerning tab issue should be fixed now and I will send you a beta.

The main problem however seems to be that you have reached the limit of how many ligatures that can be packed into a font. It looks to me like the earlier examples with fewer ligatures does work but the last two (in 1linediff) do have too many ligatures. The technical side of the problem was an integer overflow in the GSUB table when too many ligatures are processed. I have added a proper error message if this happens in the future.

**Comment by ciqven** [comment]
Hey Johan,

Thank you for the new beta, I just checked the spacing tab with some test strings and it seems to work now, good job!

About the limit of ligatures, I can see that the total amount of ligatures + contextual substitutions in the working example is 1310. I have been reading the OTF specs a bit to try and understand this, but from what I can see the tables related to GSUB all have `uint16`s for the amount, so that should be closer to `65536` entries. I also see this number mentioned on multiple online discussions when I search Google for the max amount of ligatures in OTF. From my calculations my final font should use about half of that. Is there a way to increase this limit somehow?

**Comment by johanmattssonm** [comment]
That is way too simplified, the limit in this case happens in a lookup offset. That is a pointer to some other part of the subtable. Every ligature needs coverage table entries and some extra information around it, like a feature entry, a feature entry offset, look up offset count and you also need two bytes to store the gid at any point in the table.

I am not sure what the exact limit is but it is much lower than the maximum number you can store in a uint16.

There might be a way to make Birdfont waste less space when it comes to generating ligatures but it is probably not easy. 50 bytes per ligature does not sound optimal.

**Comment by ciqven** [comment]
Thank you for the explanation, I understand now. It seems I am going to have to use classes after all to avoid needing thousands of ligatures. 

I am not sure how easy classes would be to add to Birdfont, alternatively I could export the font from Birdfont without any ligatures whatsoever. load it in FontForge and use a python script to create the tables there. Or even make my own font reader/writer. I am still not sure if what I want is possible, but Devanagari fonts seem to do it so there must be a way!

I think this issue has been completed since the Spacing tab works great now, my remaining issue is already addressed by #172 so I shall patiently wait for that to get added if/when you get around to it. Thank you for all the problem-solving so far!

**Comment by ciqven** [comment]
Hello Johan,

Good news: I got it working by exporting without ligatures from Birdfont, then adding them programmatically in Fontforge. Turns out the classes did indeed make the difference.

I'd like to thank you very much for your help figuring this out, and for making Birdfont which is simple yet complete enough that it gave me the courage to start my font project.

I am thankful for your responsiveness and commitment to this project, and have sent you a donation via the Birdfont site.

I will close this issue now, since the problem it is about (the Spacing tab) has been fixed as far as I can see. Cheers! :+1:
